### ISSUE TITLE
Duplicate symbol error when building hello.cpp

### ISSUE THREAD

**Issue by paulhuggett** [SNSystems/llvm-project-prepo#164]
Source code:
~~~cpp
#include <iostream>
int main () {
    std::cout << "Hello, World\n";
}
~~~
Follow the hello-rld/cxx build instructions [here](https://github.com/SNSystems/hello-rld/tree/main/cxx). This yields a duplicate symbol error: `_ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE5sputnEPKcl` in lib/cxx/ios.cpp.o vs. main.o. This name demangles as “std::__1::basic_streambuf<char, std::__1::char_traits<char> >::sputn(char const*, long)”.

Investigating…

~~~
$ pstore-dump --compilation=$(repo-ticket-dump lib/cxx/ios.cpp.o) clang.db  | grep -C 2 _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE5sputnEPKcl
                - digest     : 8bc7ea5a6cda8f729717ca6e209385fc
                  fext       : { addr: 7043488, size: 57 }
                  name       : _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE5sputnEPKcl
                  linkage    : weak_odr
                  visibility : default
$ pstore-dump --compilation=$(repo-ticket-dump main.o) clang.db  | grep -C 2 _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE5sputnEPKcl
                - digest     : 18a541c92b0858c39cfcb014e184e333
                  fext       : { addr: 8422240, size: 95 }
                  name       : _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE5sputnEPKcl
                  linkage    : link_once_odr
                  visibility : hidden
~~~

In the standard library, the symbol has `weak_odr` linkage. In main.o it is `link_once_odr`. The rld [symbol resolution rules](https://github.com/SNSystems/llvm-project-prepo/wiki/%5Brld%5D-Symbol-resolution-rules#the-global-symbol-table) say that this is an error, so rld is following the rules correctly. This then begs two questions:

1. Why does the symbol have different linkages in the two compilations?
2. Are the symbol resolution rules correct for this situation?

**Comment by paulhuggett** [comment]
The program can be ”disguised” by compiling main.cpp with -O3. (This eliminates the out-of-line definition of the symbol.)

**Comment by paulhuggett** [comment]
This bug was fixed by commit 2cb1101be0202e258b0da12b39d0635bfa14517c.
